src/libraries/Microsoft.Extensions.Options/src/ValidateOptionsResultBuilder.cs

```diff
@@ -0,0 +1,127 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Extensions.Options
+{
+    /// <summary>
+    /// Builds <see cref="ValidateOptionsResult"/> with support for multiple error messages.
+    /// </summary>
+    [DebuggerDisplay("{ErrorsCount} errors")]
+    public class ValidateOptionsResultBuilder
+    {
+        private const string MemberSeparatorString = ", ";
+
+        private List<string>? _errors;
+
+        /// <summary>
+        /// Creates new instance of the <see cref="ValidateOptionsResultBuilder"/> class.
+        /// </summary>
+        public ValidateOptionsResultBuilder() { }
+
+        /// <summary>
+        /// Adds validation error.
+        /// </summary>
+        /// <param name="error">Content of error message.</param>
+        /// <param name="propertyName">THe property in the option object which contains an error.</param>
+        public void AddError(string error, string? propertyName = null)
+        {
+            ThrowHelper.ThrowIfNull(error);
+            Errors.Add(propertyName is null ? error : $"Property {propertyName}: {error}");
+        }
+
+        /// <summary>
+        /// Adds any validation error carried by the <see cref="ValidationResult"/> instance to this instance.
+        /// </summary>
+        /// <param name="result">The instance to consume the errors from.</param>
```

Comment:
Nit: appears to be adding only one error and perhaps "append" is more specific than "consume" in this case.
```suggestion
        /// <param name="result">The instance to append the error from.</param>
```